src/game_events/handlers.hpp

```diff
@@ -66,17 +82,61 @@ class event_handler
 	 */
 	void handle_event(const queued_event& event_info, game_lua_kernel& lk);
 
-	const config& get_config() const
+	bool filter_event(const queued_event& event_info) const;
+
+	const config& arguments() const
+	{
+		return args_;
+	}
+
+	const std::string& id() const
+	{
+		return id_;
+	}
+
+	bool empty() const;
+
+	bool repeatable() const
+	{
+		return !first_time_only_;
+	}
+
+	// Normally non-serializable events are skipped when serializing (with a warning).
+	// If include_nonserializable is true, the game attempts to serialize them anyway.
+	// This will produce output that kind of looks like the event but would not deserialize to the same event.
+	void write_config(config& cfg, bool include_nonserializable = false) const;
+
+	void set_repeatable(bool repeat = true)
+	{
+		first_time_only_ = !repeat;
+	}
+
+	void set_menu_item(bool imi)
 	{
-		return cfg_;
+		is_menu_item_ = imi;
 	}
 
+	void set_arguments(const config& cfg)
+	{
+		args_ = cfg;
+	}
+
+	void read_filters(const config& cfg);
+	void add_filter(std::unique_ptr<event_filter>&& filter);
+
+	void register_wml_event(game_lua_kernel& lk);
+	void set_event_ref(int idx, bool has_preloaded);
+
 private:
 	bool first_time_only_;
 	bool is_menu_item_;
 	bool disabled_;
-	config cfg_;
-	std::vector<std::string> types_;
+	bool is_lua_;
```

Comment:
Documentation of these new variables would be good.

Particularly, what does `has_preloaded_` mean? AFAICS it seems to be "whether the lua kernel had already finished preloading when it last called `set_event_ref`", and I'm wondering why that previous state matters.

Reply:
It matters because events registered after preload need to be serializable in order for saved games to work, but events registered _in_ preload do _not_ need to be serializable, because the preload will re-register them when you load a saved game.

Therefore, we only issue a warning about non-serializable events if they're registered after preload.

Reply:
I think the name should be clearer about it recording the state at creation time, maybe `created_in_preload_`, but anyway this is resolved.